Fix password complexity regex for special characters (on master)

guillep2k opened the pull request:
Fixes #8521

guillep2k commented:
Also fixes the error display in the change password page.

lunny commented on modules/setting/setting.go:
```diff
@@ -785,7 +785,7 @@ func NewContext() {
 		"upper": "[A-Z]+",
 		"digit": "[0-9]+",
-		"spec":  `][ !"#$%&'()*+,./:;<=>?@\\^_{|}~` + "`-",
+		"spec":  `[][ !"#$%&'()*+,./:;<=>?@\\^_{|}~` + "`-]",
```
Maybe you should add a `+` at the end of the string.

lunny commented:
To avoid the bug again, please move `dictPC` as a gloabl variable and add a test for it.

guillep2k commented:
> 
> 
> To avoid the bug again, please move `dictPC` as a gloabl variable and add a test for it.

Will do

guillep2k commented:
@lunny @zeripath Done.

* Removed regexp completely as they complicated things unnecessarily.
* Added an 'off' setting as there was no way to disable the functionality.
* Established an `[a-zA-Z0-9]` default for password generation pattern for when complexity checks are off.
* Fixed cheat sheet and ini sample accordingly.
* Added tests.

lunny commented on modules/password/password.go:
```diff
@@ -7,45 +7,58 @@ package password
 )
 
-var matchComplexities = map[string]regexp.Regexp{}
 var matchComplexityOnce sync.Once
```
A nit thing, group all the variables will look better.
```
var (
  matchComplexityOnce sync.Once
  validChars string
)
```

guillep2k replied:
@lunny done.

lunny replied:
It seems not?

guillep2k replied:
@lunny Sorry, I had pushed to the backport branch. Now I pushed to this branch.

lunny commented on modules/password/password.go:
```diff
@@ -7,45 +7,58 @@ package password
-		for _, val := range matchComplexities {
-			if !val.MatchString(pwd) {
+	NewComplexity()
+	if len(validChars) > 0 {
```
It seems `validChars` only is used to know if we need check the complexity? How about remove `validChars` and only use `requiredChars`.

guillep2k replied:
`validChars` is used in `Generate()` to have the list of characters to choose from.
`requiredChars` is used in `IsComplexEnough()` to check the password against.

They require almost the same information, except that `Generate()` benefits from having all the characters in one string (and always requires this input) while `IsComplexEnough()` will cycle through all groups checking that the password contains at least one character from each group.
